NPE in `metabase.driver.sync/schema-patterns->filter-fn*`

(The NPE seems to be when invoking the function it returns)

Reported here on Discourse

https://discourse.metabase.com/t/snowflake-broken-in-0-48-1/63849/39

This is almost certainly causing problems with sync. We should look into the code and try to work out which situations might cause an NPE and add some safeguards against it

```
java.lang.NullPointerException
	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
	at java.base/java.util.regex.Matcher.reset(Unknown Source)
	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
	at metabase.driver.sync$schema_patterns__GT_filter_fn_STAR_$fn__80810.invoke(sync.clj:45)
	at metabase.driver.sync$include_schema_QMARK_.invokeStatic(sync.clj:76)
	at metabase.driver.sync$include_schema_QMARK_.invoke(sync.clj:66)
	at metabase.driver.snowflake$fn__128206$fn__128207$fn__128211$fn__128213.invoke(snowflake.clj:433)
	at clojure.core$filter$fn__5958$fn__5959.invoke(core.clj:2822)
	at clojure.java.jdbc$init_reduce_rs.invokeStatic(jdbc.clj:1205)
	at clojure.java.jdbc$init_reduce_rs.invoke(jdbc.clj:1197)
	at clojure.java.jdbc$reducible_result_set_STAR_$reify__28159.reduce(jdbc.clj:1227)
	at clojure.core$reduce.invokeStatic(core.clj:6885)
	at clojure.core$reduce.invoke(core.clj:6868)
	at clojure.java.jdbc$query_reducer$fn__28165$fn__28168.invoke(jdbc.clj:1262)
	at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1091)
	at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1084)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1106)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$reducible_query$reify__28182.reduce(jdbc.clj:1357)
	at clojure.core$transduce.invokeStatic(core.clj:6946)
	at clojure.core$into.invokeStatic(core.clj:6962)
	at clojure.core$into.invoke(core.clj:6950)
	at metabase.driver.snowflake$fn__128206$fn__128207$fn__128211.invoke(snowflake.clj:429)
	at metabase.driver.sql_jdbc.execute$fn__78692$fn__78693.invoke(execute.clj:388)
	at metabase.driver.sql_jdbc.execute$fn__78657$_AMPERSAND_f__78658.invoke(execute.clj:334)
	at metabase.driver.sql_jdbc.execute$fn__78657$fn__78661.invoke(execute.clj:317)
	at metabase.driver.sql_jdbc.execute$fn__78692.invokeStatic(execute.clj:382)
	at metabase.driver.sql_jdbc.execute$fn__78692.invoke(execute.clj:380)
	at clojure.lang.MultiFn.invoke(MultiFn.java:244)
	at metabase.driver.snowflake$fn__128206$fn__128207.invoke(snowflake.clj:421)
	at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:169)
	at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:150)
	at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:158)
	at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:150)
	at metabase.driver.snowflake$fn__128206.invokeStatic(snowflake.clj:418)
	at metabase.driver.snowflake$fn__128206.invoke(snowflake.clj:414)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.sync.fetch_metadata$fn__81252$_AMPERSAND_f__81253.invoke(fetch_metadata.clj:15)
	at metabase.sync.fetch_metadata$fn__81252$fn__81255.invoke(fetch_metadata.clj:12)
	at metabase.sync.sync_metadata$fn__82263$_AMPERSAND_f__82264.invoke(sync_metadata.clj:61)
	at metabase.sync.sync_metadata$fn__82263$fn__82268.invoke(sync_metadata.clj:58)
	at metabase.task.sync_databases$sync_and_analyze_database_STAR__BANG_.invokeStatic(sync_databases.clj:94)
	at metabase.task.sync_databases$sync_and_analyze_database_STAR__BANG_.invoke(sync_databases.clj:77)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invokeStatic(sync_databases.clj:113)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invoke(sync_databases.clj:101)
	at metabase.task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:118)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
[850a5ce0-29ca-4397-b800-6e8127f2a0c4] 2024-01-17T09:16:08-08:00 ERROR org.quartz.core.ErrorLogger Job (DEFAULT.metabase.task.sync-and-analyze.job threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.NullPointerException
	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
	at java.base/java.util.regex.Matcher.reset(Unknown Source)
	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
	at metabase.driver.sync$schema_patterns__GT_filter_fn_STAR_$fn__80810.invoke(sync.clj:45)
	at metabase.driver.sync$include_schema_QMARK_.invokeStatic(sync.clj:76)
	at metabase.driver.sync$include_schema_QMARK_.invoke(sync.clj:66)
	at metabase.driver.snowflake$fn__128206$fn__128207$fn__128211$fn__128213.invoke(snowflake.clj:433)
	at clojure.core$filter$fn__5958$fn__5959.invoke(core.clj:2822)
	at clojure.java.jdbc$init_reduce_rs.invokeStatic(jdbc.clj:1205)
	at clojure.java.jdbc$init_reduce_rs.invoke(jdbc.clj:1197)
	at clojure.java.jdbc$reducible_result_set_STAR_$reify__28159.reduce(jdbc.clj:1227)
	at clojure.core$reduce.invokeStatic(core.clj:6885)
	at clojure.core$reduce.invoke(core.clj:6868)
	at clojure.java.jdbc$query_reducer$fn__28165$fn__28168.invoke(jdbc.clj:1262)
	at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1091)
	at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1084)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1106)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$reducible_query$reify__28182.reduce(jdbc.clj:1357)
	at clojure.core$transduce.invokeStatic(core.clj:6946)
	at clojure.core$into.invokeStatic(core.clj:6962)
	at clojure.core$into.invoke(core.clj:6950)
	at metabase.driver.snowflake$fn__128206$fn__128207$fn__128211.invoke(snowflake.clj:429)
	at metabase.driver.sql_jdbc.execute$fn__78692$fn__78693.invoke(execute.clj:388)
	at metabase.driver.sql_jdbc.execute$fn__78657$_AMPERSAND_f__78658.invoke(execute.clj:334)
	at metabase.driver.sql_jdbc.execute$fn__78657$fn__78661.invoke(execute.clj:317)
	at metabase.driver.sql_jdbc.execute$fn__78692.invokeStatic(execute.clj:382)
	at metabase.driver.sql_jdbc.execute$fn__78692.invoke(execute.clj:380)
	at clojure.lang.MultiFn.invoke(MultiFn.java:244)
	at metabase.driver.snowflake$fn__128206$fn__128207.invoke(snowflake.clj:421)
	at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:169)
	at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:150)
	at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:158)
	at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:150)
	at metabase.driver.snowflake$fn__128206.invokeStatic(snowflake.clj:418)
	at metabase.driver.snowflake$fn__128206.invoke(snowflake.clj:414)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.sync.fetch_metadata$fn__81252$_AMPERSAND_f__81253.invoke(fetch_metadata.clj:15)
	at metabase.sync.fetch_metadata$fn__81252$fn__81255.invoke(fetch_metadata.clj:12)
	at metabase.sync.sync_metadata$fn__82263$_AMPERSAND_f__82264.invoke(sync_metadata.clj:61)
	at metabase.sync.sync_metadata$fn__82263$fn__82268.invoke(sync_metadata.clj:58)
	at metabase.task.sync_databases$sync_and_analyze_database_STAR__BANG_.invokeStatic(sync_databases.clj:94)
	at metabase.task.sync_databases$sync_and_analyze_database_STAR__BANG_.invoke(sync_databases.clj:77)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invokeStatic(sync_databases.clj:113)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invoke(sync_databases.clj:101)
	at metabase.task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:118)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
```

--- a/src/metabase/driver/sync.clj
+++ b/src/metabase/driver/sync.clj
@@ -40,11 +40,16 @@
 
       :else
       (let [inclusion? exclusion-blank?
-            pattern    (schema-pattern->re-pattern (if inclusion? inclusion-patterns exclusion-patterns))]
-        (fn [s]
-          (let [m        (.matcher pattern s)
-                matches? (.matches m)]
-            (if inclusion? matches? (not matches?))))))))
+            pattern    (schema-pattern->re-pattern (if inclusion? inclusion-patterns exclusion-patterns))
+            match-fn   (fn match-fn [^String s]
+                         (when s
+                           (let [m (.matcher pattern s)]
+                             (.matches m))))]
+        ;; for inclusion patterns, never match a `nil` schema; for exclusion patterns, always consider a `nil` schema to
+        ;; be ok
+        (if inclusion?
+          match-fn
+          (complement match-fn))))))
 
 (def ^:private schema-patterns->filter-fn (memoize schema-patterns->filter-fn*))
 
@@ -56,12 +61,13 @@
    (let [{prop-name :name} (driver.u/find-schema-filters-prop (driver.u/database->driver database))]
      (db-details->schema-filter-patterns prop-name database)))
   ([prop-nm {db-details :details :as _database}]
-   (let [schema-filter-type     (get db-details (keyword (str prop-nm "-type")))
-         schema-filter-patterns (get db-details (keyword (str prop-nm "-patterns")))]
-     (case schema-filter-type
-       "exclusion" [nil schema-filter-patterns]
-       "inclusion" [schema-filter-patterns nil]
-       [nil nil]))))
+   (when prop-nm
+     (let [schema-filter-type     (get db-details (keyword (str prop-nm "-type")))
+           schema-filter-patterns (get db-details (keyword (str prop-nm "-patterns")))]
+       (case schema-filter-type
+         "exclusion" [nil schema-filter-patterns]
+         "inclusion" [schema-filter-patterns nil]
+         [nil nil])))))
 
 (defn include-schema?
   "Returns true if the given `schema-name` should be included/synced, considering the given `inclusion-patterns` and

--- a/test/metabase/driver/sync_test.clj
+++ b/test/metabase/driver/sync_test.clj
@@ -1,9 +1,12 @@
 (ns metabase.driver.sync-test
   (:require
    [clojure.test :refer :all]
+   [metabase.driver.postgres]
    [metabase.driver.sync :as driver.s])
   (:import
    (clojure.lang ExceptionInfo)))
+
+(comment metabase.driver.postgres/keep-me) ; this is used by [[schema-filter-NPE-test]] below
 
 (deftest ^:parallel schema-filter-test
   (doseq [[test-kind expect-match? schema-name inclusion-filters exclusion-filters]

--- a/test/metabase/driver/sync_test.clj
+++ b/test/metabase/driver/sync_test.clj
@@ -1,11 +1,11 @@
 (ns metabase.driver.sync-test
   (:require
-   [clojure.test :as t]
+   [clojure.test :refer :all]
    [metabase.driver.sync :as driver.s])
   (:import
    (clojure.lang ExceptionInfo)))
 
-(t/deftest ^:parallel schema-filter-test
+(deftest ^:parallel schema-filter-test
   (doseq [[test-kind expect-match? schema-name inclusion-filters exclusion-filters]
           [["nil filters" true "foo" nil nil]
            ["blank filters" true "foo" "" ""]
@@ -20,10 +20,23 @@
            ["exclusion filter with commas and wildcards (include)" true "foo" "" "ba*,fob"]
            ["exclusion filter with commas and wildcards (exclude)" false "foo" "" "bar,baz,fo*"]
            ["multiple inclusion with whitespace trimming" true "bar" "  foo  ,  bar \n  ,  \nbaz  " ""]]]
-    (t/testing (str "include-schema? works as expected for " test-kind)
-      (t/is (= expect-match? (driver.s/include-schema? inclusion-filters exclusion-filters schema-name))))
-    (t/testing "include-schema? throws an exception if both patterns are specified"
-      (t/is (thrown-with-msg?
-             ExceptionInfo
-             #"Inclusion and exclusion patterns cannot both be specified"
-             (driver.s/include-schema? "foo" "bar" "whatever"))))))
+    (testing (str "include-schema? works as expected for " test-kind)
+      (is (= expect-match? (driver.s/include-schema? inclusion-filters exclusion-filters schema-name))))
+    (testing "include-schema? throws an exception if both patterns are specified"
+      (is (thrown-with-msg?
+           ExceptionInfo
+           #"Inclusion and exclusion patterns cannot both be specified"
+           (driver.s/include-schema? "foo" "bar" "whatever"))))))
+
+(deftest ^:parallel schema-filter-NPE-test
+  (testing "Schema filter function should not NPE if you pass in a `nil` schema (#38156)"
+    (testing "inclusion -- don't include nil schemas"
+      (let [db {:details {:schema-filters-type "inclusion"
+                          :schema-filters-patterns "x"}
+                :engine :postgres}]
+        (is (not (driver.s/include-schema? db nil)))))
+    (testing "exclusion -- don't exclude nil schemas"
+      (let [db {:details {:schema-filters-type "exclusion"
+                          :schema-filters-patterns "x"}
+                :engine :postgres}]
+        (is (driver.s/include-schema? db nil))))))

